content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md

```diff
@@ -0,0 +1,146 @@
+| `self >= oldSelf` | Validate that a number is monotonically increasing.|
+
+
+## Using the Functions Libraries
+
+Validation rules have access to a couple different function libraries:
+
+* CEL standard functions, defined in the [list of standard definitions](https://github.com/google/cel-spec/blob/v0.7.0/doc/langdef.md#list-of-standard-definitions)
+* CEL standard [macros](https://github.com/google/cel-spec/blob/v0.7.0/doc/langdef.md#macros)
+* CEL [extended string function library](https://pkg.go.dev/github.com/google/cel-go@v0.11.2/ext#Strings)
+* Kubernetes [CEL extension library](https://pkg.go.dev/k8s.io/apiextensions-apiserver@v0.24.0/pkg/apiserver/schema/cel/library#pkg-functions) which includes supplemental functions for [lists](https://pkg.go.dev/k8s.io/apiextensions-apiserver@v0.24.0/pkg/apiserver/schema/cel/library#pkg-functions), [regex](https://pkg.go.dev/k8s.io/apiextensions-apiserver@v0.24.0/pkg/apiserver/schema/cel/library#Regex), and [URLs](https://pkg.go.dev/k8s.io/apiextensions-apiserver@v0.24.0/pkg/apiserver/schema/cel/library#Regex).
+
+Examples of function libraries in use:
+
+| Validation Rule | Purpose |
+| - | - |
+| `!(self.getDayOfWeek() in [0, 6])` | Validate that a date is not a Sunday or Saturday. |
+| `isUrl(self) && url(self).getHostname() in [a.example.com', 'b.example.com']` | Validate that a URL has an allowed hostname. |
+| `self.map(x, x.weight).sum() == 1` | Validate that the weights of a list of objects sum to 1. |
+| `int(self.find('^[0-9]*')) < 100` | Validate that a string starts with a number less than 100. |
+| `self.isSorted()` | Validates that a list is sorted. |
+
+## Resource Limits
+
+To prevent CEL evaluation from consuming excessive compute resources, validation rules impose some limits. These limits are based on CEL "cost units", a platform and machine independent measure of execution cost. As a result, the limits are the same regardless of where they are enforced.
+
+### Estimated Cost Limit
+
+CEL is, by design, non-Turing-complete in such a way that the halting problem isn’t a concern. CEL takes advantage of this design choice to include an "estimated cost" subsystem that can statically compute the worst case run time cost of any CEL expression. Validation rules are [integrated with the estimated cost system](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#resource-use-by-validation-functions) and disallow CEL expressions from being included in CRDs if they have a sufficiently poor (high) estimated cost. The estimated cost limit is set quite high and typically requires an O(n^2) or worse operation, across something of unbounded size, to be exceeded. Fortunately the fix is usually quite simple: because the cost system is aware of size limits declared in the CRD's schema, CRD authors can add size limits to the CRD's schema  (`maxItems` for arrays, `maxProperties` for maps, `maxLength` for strings) to reduce the estimated cost.
```

Comment:
```suggestion
CEL is, by design, non-Turing-complete in such a way that the halting problem isn’t a concern. CEL takes advantage of this design choice to include an "estimated cost" subsystem that can statically compute the worst case run time cost of any CEL expression. Validation rules are [integrated with the estimated cost system](o/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#resource-use-by-validation-functions) and disallow CEL expressions from being included in CRDs if they have a sufficiently poor (high) estimated cost. The estimated cost limit is set quite high and typically requires an O(n^2) or worse operation, across something of unbounded size, to be exceeded. Fortunately the fix is usually quite simple: because the cost system is aware of size limits declared in the CRD's schema, CRD authors can add size limits to the CRD's schema  (`maxItems` for arrays, `maxProperties` for maps, `maxLength` for strings) to reduce the estimated cost.
```

_aside_
The docs for the cost estimation system really belong within either https://kubernetes.io/docs/reference/ or https://kubernetes.io/docs/concepts/
Task pages are really for explaining what the reader must do to achieve an outcome, and therefore aren't the right place for longer explanations.

Reply:
This is a really good point.  I feel like we've got a more general problem with CRD documentation here. I'll track it and start working on reorganizing it to fit the concept/task grouping.